nomt/src/beatree/ops/update/branch_ops.rs

```diff
@@ -0,0 +1,965 @@
+                start: chunk.start,
+                end: chunk.start + left_chunk_n_items,
+                sum_separator_lengths: left_chunk_sum_separator_lengths,
+            };
+
+            let right_chunk = KeepChunk {
+                start: chunk.start + left_chunk_n_items,
+                end: chunk.end,
+                sum_separator_lengths: chunk.sum_separator_lengths
+                    - left_chunk_sum_separator_lengths,
+            };
+
+            self.ops.insert(index, BranchOp::KeepChunk(left_chunk));
+            self.ops[index + 1] = BranchOp::KeepChunk(right_chunk);
+        }
+        left_chunk_n_items
+    }
+}
+
+pub struct BranchOps<'a> {
+    ops: &'a mut Vec<BranchOp>,
+    end: usize,
+}
+
+impl<'a> Deref for BranchOps<'a> {
+    type Target = [BranchOp];
+    fn deref(&self) -> &Self::Target {
+        &self.ops[..self.end]
+    }
+}
+
+impl<'a> Drop for BranchOps<'a> {
+    fn drop(&mut self) {
+        self.ops.drain(..self.end);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::beatree::{
+        branch::node::{self, get_key, BRANCH_NODE_BODY_SIZE},
+        ops::{
+            bit_ops::separator_len,
+            update::{
+                branch_ops::{BranchOp, BranchOpsTracker, KeepChunk},
+                branch_updater::{
+                    tests::{make_branch, make_branch_with_body_size_target, prefixed_key},
+                    BaseBranch, BranchGauge,
+                },
+                BRANCH_BULK_SPLIT_TARGET, BRANCH_BULK_SPLIT_THRESHOLD, BRANCH_MERGE_THRESHOLD,
+            },
+        },
+        Key, PageNumber,
+    };
+
+    #[test]
+    fn bulk_split() {
+        let key = |i| prefixed_key(0x00, 16, i);
+
+        let mut gauge = BranchGauge::default();
+        let mut n_keys = 0;
+
+        let mut ops_tracker = BranchOpsTracker::new();
+
+        while gauge.body_size() < BRANCH_BULK_SPLIT_THRESHOLD {
+            let key = key(n_keys);
+
+            gauge.ingest_key(key, separator_len(&key));
+            ops_tracker.push_insert(key, PageNumber(n_keys as u32));
+
+            n_keys += 1;
+        }
+
+        let mut n_splits = 0;
+        let mut n_items_in_splits = 0;
+        while let Some((ops, _gauge)) =
+            ops_tracker.extract_ops_until(None, BRANCH_BULK_SPLIT_TARGET)
+        {
+            n_splits += 1;
+            n_items_in_splits += ops.len();
+        }
+
+        let expected_bulk_split = BRANCH_BULK_SPLIT_THRESHOLD / BRANCH_BULK_SPLIT_TARGET;
+
+        // After ingesting all those ops, the bulk split is expected to be initiated.
+        assert_eq!(n_splits, expected_bulk_split);
+        // The gauge needs to contain all the ops not present in the previous split.
+        assert_eq!(ops_tracker.gauge.n(), n_keys - n_items_in_splits);
+        // Operations are expected to be drainded.
+        assert_eq!(ops_tracker.ops.len(), n_keys - n_items_in_splits);
+    }
+
+    #[test]
+    fn exrtact_ops_until_only_inserts() {
```

Comment:
typo here and elsewhere `exrtact`

Reply:
Weird. I'm using `cargo-spellcheck` to ensure I'm not leaving typos or more obvious errors, but it didn't flag those. I will correct them now obv!